### PR TITLE
[ModLog] Optimise get_next_case_number()

### PR DESCRIPTION
**Pull request by Tobotimus**
We've noticed that this is taking an obscenely long time on very large servers.

### CODE REVIEW

File: redbot/core/modlog.py [hunk 1]
```diff
@@ -174,6 +174,14 @@
...
         await _conf.schema_version.set(3)
 
+    if from_version < 4 <= to_version:
+        # set latest_case_number
+        for guild_id, cases in (await _conf.custom(_CASES).all()).items():
+            if cases:
+                await _conf.guild(
+                    cast(discord.Guild, discord.Object(id=guild_id))
+                ).latest_case_number.set(max(map(int, cases.keys())))
+
```
Comment (mikeshardmind) [inline redbot/core/modlog.py:183]: you need to set the schema version to 4 in this block

**Review by mikeshardmind — CHANGES_REQUESTED** [review]

File: redbot/core/modlog.py (hunk 1 above)
Comment (Tobotimus) [reply redbot/core/modlog.py:183]: Good point, forgot about that :+1:

File: redbot/core/modlog.py [hunk 2]
```diff
@@ -763,6 +767,7 @@ def from_json(cls, name: str, data: dict, **kwargs):
         message=None,
     )
     await _conf.custom(_CASES, str(guild.id), str(next_case_number)).set(case.to_json())
```
Comment (mikeshardmind) [inline redbot/core/modlog.py]: This needs to be inside a lock on the value for case numbers. We are fetching, awaiting something else, then incrementing here.

**Review by mikeshardmind — DISMISSED** [review]
found something else...

File: redbot/core/modlog.py (hunk 2 above)
Comment (Tobotimus) [reply redbot/core/modlog.py]: Good point

**Review by mikeshardmind — CHANGES_REQUESTED** [review]
one last change from me after some consideration.

Let's go ahead and make `get_next_case_number` private and remove it from `__all__`

There's not a reason to expose this function outside of modlog, and the next release is one which allows for removing this from the public API.

**Comment by Tobotimus** [comment]
@mikeshardmind I've actually gone ahead and removed the function all together. It was only used in one place, and that place is kind of easier to understand (w.r.t. lock usage) if it just uses config directly.

**Review by mikeshardmind — DISMISSED** [review]

**Review by mikeshardmind — APPROVED** [review]

**Pull request merged**